packages/url_launcher/url_launcher_macos/example/macos/RunnerTests/RunnerTests.swift

```diff
@@ -6,19 +6,105 @@ import FlutterMacOS
 import XCTest
 import url_launcher_macos
 
+/// A stub to simulate the system Url handler.
+class StubWorkspace: SystemURLHandler {
+
+  var isSuccessful = true
+
+  func open(_ url: URL) -> Bool {
+    return isSuccessful
+  }
+
+  func urlForApplication(toOpen: URL) -> URL? {
+    return toOpen
+  }
+}
+
 class RunnerTests: XCTestCase {
-  func testCanLaunch() throws {
+
+  func testNoHandlerReturnsFlutterMethodNotImplemented() throws {
+    let plugin = UrlLauncherPlugin()
+
+    let call = FlutterMethodCall(
+      methodName: "noHandler",
+      arguments: ["url": "https://flutter.dev"])
+
+    plugin.handle(
+      call,
+      result: { (result: Any?) -> Void in
+        XCTAssertEqual(result as? NSObject, FlutterMethodNotImplemented)
+      })
+  }
+
+  func testCanLaunchSuccessReturnsTrue() throws {
     let plugin = UrlLauncherPlugin()
+
     let call = FlutterMethodCall(
       methodName: "canLaunch",
       arguments: ["url": "https://flutter.dev"])
-    var canLaunch: Bool?
+
+    plugin.handle(
+      call,
+      result: { (result: Any?) -> Void in
+        XCTAssertEqual(result as? Bool, true)
+      })
+  }
+
+  func testCanLaunchInvalidUrlReturnsFalse() throws {
+    let plugin = UrlLauncherPlugin()
+
+    let call = FlutterMethodCall(
+      methodName: "canLaunch",
+      arguments: ["url": "brokenUrl"])
+
     plugin.handle(
       call,
       result: { (result: Any?) -> Void in
-        canLaunch = result as? Bool
+        XCTAssertEqual(result as? Bool, false)
+      })
+  }
+
+  func testCanLaunchMissingArgumentReturnsFlutterError() throws {
+    let plugin = UrlLauncherPlugin()
+
+    let call = FlutterMethodCall(
+      methodName: "canLaunch",
+      arguments: [])
+
+    plugin.handle(
+      call,
+      result: { (result: Any?) -> Void in
+        XCTAssertTrue(result is FlutterError)
+      })
+  }
+
+  func testLaunchSuccessReturnsTrue() throws {
+    let workspace = StubWorkspace()
+    let pluginWithStubWorkspace = UrlLauncherPlugin(workspace)
+
+    let call = FlutterMethodCall(
+      methodName: "launch",
+      arguments: ["url": "https://flutter.dev"])
+
+    pluginWithStubWorkspace.handle(
+      call,
+      result: { (result: Any?) -> Void in
+        XCTAssertEqual(result as? Bool, true)
       })
+  }
 
```

Comment:
Similarly, `launch` can fail in exactly the same way. You've set up the mock handler to have the ability to return false, but you're not ever using it to write a test for launch failure.

Reply:
Thank you, now the stub is in use for testing when `launch` cannot open the URL.